File: backend/src/services/transcriber.ts
import { spawn } from "node:child_process";
import { createReadStream } from "node:fs";
import { unlink } from "node:fs/promises";
import OpenAI from "openai";
import type { Transcript, WordTimestamp } from "../types/index.js";

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

function extractAudio(videoPath: string, audioPath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const proc = spawn(process.env.FFMPEG_PATH ?? "/opt/homebrew/bin/ffmpeg", [
      "-i",
      videoPath,
      "-vn",
      "-ar",
      "16000",
      "-ac",
      "1",
      "-b:a",
      "64k",
      "-y",
      audioPath,
    ]);

    let stderr = "";
    proc.stderr.on("data", (d: Buffer) => {
      stderr += d.toString();
    });

    proc.on("close", (code) => {
      if (code === 0) resolve();
      else reject(new Error(`ffmpeg exited with code ${code}: ${stderr}`));
    });

    proc.on("error", (err) =>
      reject(new Error(`Failed to spawn ffmpeg: ${err.message}`)),
    );
  });
}

export async function transcribeVideo(videoPath: string): Promise<Transcript> {
  const audioPath = videoPath.replace(/\.[^.]+$/, ".mp3");

  await extractAudio(videoPath, audioPath);

  try {
    const response = await openai.audio.transcriptions.create({
      file: createReadStream(audioPath),
      model: "whisper-1",
      response_format: "verbose_json",
      timestamp_granularities: ["word"],
    });

    const words: WordTimestamp[] = (response.words ?? []).map((w) => ({
      word: w.word,
      start: w.start,
      end: w.end,
    }));

    return { text: response.text, words };
  } finally {
    await unlink(audioPath).catch(() => {});
  }
}

File: backend-python/services/transcriber.py
import os
import subprocess
import re
from openai import OpenAI
from models import Transcript, WordTimestamp

_client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))
_FFMPEG = os.environ.get("FFMPEG_PATH", "/opt/homebrew/bin/ffmpeg")


def _extract_audio(video_path: str, audio_path: str) -> None:
    result = subprocess.run(
        [_FFMPEG, "-i", video_path, "-vn", "-ar", "16000", "-ac", "1", "-b:a", "64k", "-y", audio_path],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(f"ffmpeg audio extract failed: {result.stderr}")


def transcribe_video(video_path: str) -> Transcript:
    audio_path = re.sub(r"\.[^.]+$", ".mp3", video_path)
    _extract_audio(video_path, audio_path)

    try:
        with open(audio_path, "rb") as f:
            response = _client.audio.transcriptions.create(
                file=f,
                model="whisper-1",
                response_format="verbose_json",
                timestamp_granularities=["word"],
            )

        words = [
            WordTimestamp(word=w.word, start=w.start, end=w.end)
            for w in (response.words or [])
        ]
        return Transcript(text=response.text, words=words)
    finally:
        if os.path.exists(audio_path):
            os.unlink(audio_path)
